### ISSUE TITLE
Using a yarn workspace, my sub-packages are missing all dependencies except for those declared in the top-level package.json

### ISSUE THREAD

**Issue by ZempTime** [lerna/lerna#1561]
We're getting moved into a monorepo. I've added the `lerna.json` and our top-level `package.json` to the details at the bottom of this post.

## Current Behavior
When I run:
```
lerna clean --yes
yarn install OR lerna bootstrap
```
Nothing blows up. It looks like everything worked. But then when I hop inside one of my sub-packages and try to boot it up, I get:

<img width="1277" alt="screen shot 2018-08-13 at 10 06 01 am" src="https://user-images.githubusercontent.com/3596468/44040289-8399b186-9ee0-11e8-8aaa-214b2e8be911.png">

When I hop inside my `packages/sub-package/node_modules`, I only see the dependencies declared in the top-level `package.json`, and no symlinks or anything of the sort.

## Expected Behavior
I run some series of commands. All of the dependencies install so I can both:

* have `lerna run lint` actually execute across all sub-packages
* be able to start up my UI

At this point, my assumption is that I'm doing something wrong and the tools are working correctly. But I'm not sure how to dig into this further.

## My Questions

* What's the standard operating procedure here such that this _should_ work?
* What's the next most valuable piece of information I can go collect & put back here?

<details><summary>lerna.json</summary><p>

```json
{
  "lerna": "2.11.0",
  "packages": [
    "packages/*"
  ],
  "bootstrap": {
    "ignore": "*"
  },
  "useWorkspaces": true,
  "npmClient": "yarn",
  "version": "0.0.1"
}
```
</p></details>

<details><summary>top-level package.json</summary><p>

```
{
  "watch": {
    "update-styling": {
      "patterns": [
        "packages"
      ],
      "extensions": "scss"
    }
  },
  "scripts": {
    "installDev": "./scripts/dev-install.sh",
    "installProd": "./scripts/prod-install.sh",
    "refresh": "./scripts/dev-refresh.sh",
    "dev": "./scripts/dev-servers-start.sh",
    "update-styling": "./scripts/update-styling.sh",
    "watch": "npm-watch",
    "compile": "lerna run compile",
    "clean": "lerna run clean",
    "lint-all": "lerna run lint",
    "update-version": "lerna publish --skip-git --skip-npm --force-publish '*'",
    "check-packages": "yarn clean && yarn compile && yarn test && yarn lint",
    "publish-packages": "./scripts/publish"
  },
  "devDependencies": {
    "async": "^2.6.1",
    "babel-eslint": "^8.2.6",
    "eslint": "^5.3.0",
    "eslint-config-google": "^0.9.1",
    "eslint-plugin-babel": "^5.1.0",
    "eslint-plugin-chai-expect": "^1.1.1",
    "eslint-plugin-html": "^4.0.5",
    "eslint-plugin-maritz": "^0.3.0",
    "lerna": "^2.11.0",
    "npm-watch": "^0.3.0",
    "pre-push": "^0.1.1"
  },
  "workspaces": [
    "packages/*"
  ],
  "private": true,
  "pre-push": [
    "lint-all"
  ]
}
```
</p></details>

<details><summary>packages/singularity-ui/package.json</summary><p>

```json
{
  "name": "@maritz/singularity-ui",
  "version": "0.0.1",
  "contributors": [
    "Maritz"
  ],
  "repository": "hive-network/mz-singularity",
  "watch": {
    "update-styling": "../*/{src}/*.scss"
  },
  "scripts": {
    "start": "polymer serve",
    "build": "npm run build:prpl-server && npm run build:static",
    "build:prpl-server": "gulp clean:prpl-server && gulp build:prpl-server",
    "build:static": "polymer build",
    "serve:prpl-server": "prpl-server --root server/build",
    "serve:static": "polymer serve --port 5000 build/es5-bundled",
    "test": "npm run test:unit && npm run test:integration",
    "test:integration": "mocha test/integration --timeout=10000",
    "test:unit": "polymer test --module-resolution=node --npm",
    "test:sauce": "polymer test --module-resolution=node --npm -s 'windows 10/microsoftedge@17' -s 'macos 10.13/safari@11'",
    "test:regenerate_screenshots": "mocha test/integration/screenshots-baseline/regenerate.js --timeout=15000",
    "dev": "polymer serve --port 8900 | npm run watch-files",
    "watch-files": "browser-sync start --proxy 127.0.0.1:8900 --reload-delay 20000 --files src/*.js, images/* | npm run watch",
    "watch": "cd ../../; npm run watch",
    "lint": "npm run lint:javascript && polymer lint",
    "lint:javascript": "eslint . --ext js",
    "update-styling": "cd ../../;scripts/update-styling.sh"
  },
  "dependencies": {
    "@maritz/mtztools-graphql-helpers": "^0.0.1",
    "@maritz/mtzwc-theme": "^0.0.1",
    "@maritz/mtzwc-nav": "^0.0.1",
    "@maritz/mtzwc-program-card": "^0.0.1",
    "@maritz/mtzwc-user-avatar": "^0.0.1",
    "@maritz/mtzwc-user-nav": "^0.0.1",
    "@maritz/singularity-auth-ui": "^0.0.1",
    "@maritz/singularity-config-ui": "^0.0.1",
    "@maritz/singularity-example-ui": "^0.0.1",
    "@maritz/singularity-header-ui": "^0.0.1",
    "@maritz/singularity-landing-ui": "^0.0.1",
    "@maritz/singularity-store-ui": "^0.0.1",
    "@maritz/singularity-theme-ui": "^0.0.1",
    "@material/form-field": "^0.35.0",
    "@material/mwc-button": "^0.1.2",
    "@material/mwc-fab": "^0.1.2",
    "@material/mwc-icon": "^0.1.2",
    "@material/textfield": "^0.35.2",
    "@polymer/app-layout": "^3.0.0-pre.19",
    "@polymer/iron-flex-layout": "^3.0.0-pre.15",
    "@polymer/lit-element": "^0.5.0",
    "@webcomponents/webcomponentsjs": "^2.0.0",
    "pwa-helpers": "^0.8.2",
    "redux": "^3.7.2",
    "redux-thunk": "^2.2.0",
    "reselect": "^3.0.1"
  },
  "devDependencies": {
    "axe-core": "^3.0.0",
    "babel-eslint": "^8.2.6",
    "browser-sync": "^2.24.4",
    "chai": "^4.1.2",
    "del": "^3.0.0",
    "eslint": "^5.2.0",
    "gulp": "^4.0.0",
    "mocha": "^5.1.0",
    "npm-watch": "^0.3.0",
    "pixelmatch": "^4.0.2",
    "polymer-cli": "^1.7.1",
    "polyserve": "^0.27.0",
    "prpl-server": "^1.1.0",
    "puppeteer": "^1.4.0",
    "renamer": "^0.6.1",
    "replace": "^1.0.0",
    "wct-browser-legacy": "^1.0.1"
  },
  "private": true
}
```
</p></details>

## Your Environment

| Executable | Version |
| ---: | :--- |
| `lerna --version` | 3.0.0-rc.0 |
| `npm --version`  | 5.6.0 |
| `yarn --version` | 1.9.4 |
| `node --version` | v9.10.0 |

| OS | Version |
| --- | --- |
| MacOS High Sierra | 10.13.4 |

**Comment by ZempTime** [comment]
I think this might be a duplicate of https://github.com/lerna/lerna/issues/1330

**Comment by evocateur** [comment]
In a lerna monorepo, your build + test scripts should generally be run from the root, as that is the location you should spend most of your time at in the shell during development.

Running linters from the root is (relatively) trivial. I have no experience with Polymer, but maybe that's not a good fit for a lerna monorepo right now?

As explained in the linked issue, when using Yarn workspaces `lerna bootstrap` is effectively the same as `yarn install`, and can largely be removed from the development cycle.

**Comment by ZempTime** [comment]
> In a lerna monorepo, your build + test scripts should generally be run from the root, as that is the location you should spend most of your time at in the shell during development.

@evocateur, thank you for making me think I'm not crazy. As I've been staring at this and learning about yarn workspaces I'm wondering why that's not the case currently

I know people are [doing it](https://github.com/slanted/polymer3-example) (theres an example). I'm just now getting into a codebase and finding I can't both lint and serve the ui without a bunch of extra bash scripts and crazy folder sizes. We can _probably_ consider this issue closed, although I'm gonna come back here and report what got me to 'working' regardless of whether or not the issue is open

**Comment by evocateur** [comment]
@ZempTime Thanks in advance for following up :)

**Comment by ZempTime** [comment]
I got this working. I realized in https://github.com/Polymer/tools/issues/273 I could just symlink across my packages up to the top-level `node_modules`. Not sure why I didn't just do this in the first place.

As for _how_ I did it, here's the api surface I built around. I added this to my root `package.json`:

```
    "symlinks:apply": "node ./packages/monorepo-cli/bin/symlinks.js apply",
    "symlinks:remove": "node ./packages/monorepo-cli/bin/symlinks.js remove",
```

I have a bunch of convenient scripts and things very specific to my project, so I've been gathering those under a `monorepo-cli` package. This act of symlinking the way I need became one of those things.

Now I just run `yarn run symlinks:apply` & layer that into other commands, like so:

```
    "monorepo:setup": "yarn install && yarn run symlinks:apply",
```

To make sure I was forward-compatible with changes I couldn't anticipate in the project, I've added a  flag to all the `package.json`'s:

```
"myOrg": {
  "symlinkWorkspaces": true
}
```

If that's on, I symlink. So I'm down to a one command to get up and running, even on new machines. And it's beautiful :)

Closing this issue as it's now resolved
